bokeh/charts/_builder.py

```diff
@@ -76,8 +81,26 @@ class Builder(HasProps):
 
     x_range = Instance(Range)
     y_range = Instance(Range)
+    y_names = Seq(String)
+    x_names = Seq(String)
```

Comment:
Image data has a two dimensional index, it already don't fit in this scheme by definition. Another way to describe things: it's array oriented, not table/column oriented. Anything graph-related will not fit in this scheme either. My suggestion is not a radical change, it is about "leaving room" in the future for things that are charts, but aren't these kind of charts. Otherwise the future will bring either i) hacks and kludges, ii) big refactors, or iii) deciding not to do something we otherwise would want to. I also like injecting an additional class in this instance for the additional reason that I very strongly favor having the `Builder` base class specify the interface that all builders must support, *and not do anything else*. Basically I think `Builder` should be an abstract base class with no implementation to speak of. That will help with documentation. Just to be clear, this is all I am suggesting:
```
class Builder:
    # unimplemented public methods that all builders must provide
    # possibly some private helper functions actually implemented

class TabularSourceBuilder
    # all the implementation for the xnames, ynames functionality goes here 
    # added bonus: the `__init__` signature can actually have xnames and ynames as real
    # parameters! Much better for docs

class TimeSeriesBuilder(TabularSourceBuilder):
    # time series builder implementation
```
Now when we need `GraphSourceBuilder` or `ArraySourceBuilder` or `MapSourceBuilder`, there is a place they can go without having to re-work everything.